RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc

```diff
@@ -0,0 +1,628 @@
+         throw cms::Exception("PixelCPEClusterRepair")
+	 << "\nERROR: Template ID " << barrelTemplateID_ << " not loaded correctly from text file. Reconstruction will fail.\n\n";
+      
+      if ( !SiPixelTemplate::pushfile( forwardTemplateID_ , thePixelTemp_ , templateDir_ ) )
+         throw cms::Exception("PixelCPEClusterRepair")
+	 << "\nERROR: Template ID " << forwardTemplateID_ << " not loaded correctly from text file. Reconstruction will fail.\n\n";
+   }
+   
+   speed_ = conf.getParameter<int>( "speed");
+   LogDebug("PixelCPEClusterRepair::PixelCPEClusterRepair:") <<
+   "Template speed = " << speed_ << "\n";
+   
+   UseClusterSplitter_ = conf.getParameter<bool>("UseClusterSplitter");   
+
+
+   //--- Configure 3D reco.
+   if ( conf.exists("MinProbY") )
+     minProbY_ = conf.getParameter<double>("MinProbY");
+   else
+     minProbY_ = 0.001;           // probabilityY < 0.001
+
+   if ( conf.exists("MaxSizeMismatchInY") )
+     maxSizeMismatchInY_ = conf.getParameter<int>("MaxSizeMismatchInY");
+   else
+     maxSizeMismatchInY_ = 1;     // ( templ.clsleny() - nypix > 1)
+
+}
+
+
+
+//-----------------------------------------------------------------------------
+//  Clean up.
+//-----------------------------------------------------------------------------
+PixelCPEClusterRepair::~PixelCPEClusterRepair()
+{
+   for(auto x : thePixelTemp_) x.destroy();
+   // Note: this is not needed for Template 2D
+}
+
+PixelCPEBase::ClusterParam* PixelCPEClusterRepair::createClusterParam(const SiPixelCluster & cl) const
+{
+   return new ClusterParamTemplate(cl);
+}
+
+
+
+//------------------------------------------------------------------
+//  Public methods mandated by the base class.
+//------------------------------------------------------------------
+
+//------------------------------------------------------------------
+//  The main call to the template code.
+//------------------------------------------------------------------
+LocalPoint
+PixelCPEClusterRepair::localPosition(DetParam const & theDetParam, ClusterParam & theClusterParamBase) const
+{
+   
+   ClusterParamTemplate & theClusterParam = static_cast<ClusterParamTemplate &>(theClusterParamBase);
+   
+   if(!GeomDetEnumerators::isTrackerPixel(theDetParam.thePart))
+      throw cms::Exception("PixelCPEClusterRepair::localPosition :")
+      << "A non-pixel detector type in here?";
+   
+   int ID = -9999;
+   if ( LoadTemplatesFromDB_ ) {
+      int ID0 = templateDBobject_->getTemplateID(theDetParam.theDet->geographicalId()); // just to comapre
+      ID = theDetParam.detTemplateId;
+      if(ID0!=ID) cout<<" different id"<< ID<<" "<<ID0<<endl;
+   } else { // from asci file
+      if ( ! GeomDetEnumerators::isEndcap(theDetParam.thePart) )
+	ID = barrelTemplateID_  ; // barrel
+      else
+	ID = forwardTemplateID_ ; // forward
+   }
+   //cout << "PixelCPEClusterRepair : ID = " << ID << endl;
```

Comment:
Please remove the commented out line, or replace with some LogDebug or similar

Reply:
Same as above.  This is a vital information for establishing we are running the correct templates.